src/net/tcp/nbd.c

```diff
@@ -0,0 +1,1095 @@
+		nbd->tx_state = NBD_TX_IDLE;
+
+	return 0;
+}
+
+/**
+ * NBD transmit process
+ *
+ * @v nbd		NBD session
+ */
+static void nbd_tx_step ( struct nbd_session *nbd ) {
+	int ( * tx ) ( struct nbd_session *nbd );
+	int rc;
+
+	while ( 1 ) {
+		DBGCIO ( nbd, "NBD %p try to transmit from state %d\n",
+			 nbd, nbd->tx_state );
+
+		switch ( nbd->tx_state ) {
+		case NBD_TX_IDLE:
+			/* Nothing to do; pause processing */
+			nbd_tx_pause ( nbd );
+			return;
+
+		case NBD_TX_NEG_OPT:
+			tx = nbd_tx_neg_opt;
+			break;
+
+		case NBD_TX_CMD_BRC:
+			/* block_read_capacity command don't transmit really */
+			nbd_tx_cmd_brc ( nbd );
+			continue;
+
+		case NBD_TX_CMD_HEADER:
+			tx = nbd_tx_cmd_header;
+			break;
+
+		case NBD_TX_CMD_DATA:
+			tx = nbd_tx_cmd_data;
+			break;
+
+		default:
+			assert ( 0 );
+			return;
+		}
+
+		/* Check for window availability */
+		if ( xfer_window ( &nbd->socket ) == 0 ) {
+			/* Cannot transmit at this point; pause
+			 * processing and wait for window to reopen
+			 */
+			nbd_tx_pause ( nbd );
+			return;
+		}
+
+		/* Transmit data */
+		if ( ( rc = tx ( nbd ) ) != 0 ) {
+			DBGC ( nbd, "NBD %p could not transmit: %s\n",
+			       nbd, strerror ( rc ) );
+			/* Transmission errors are fatal */
+			nbd_close ( nbd, rc );
+			return;
+		}
+	}
+}
+
+
+/** NBD command reply process descriptor */
+static struct process_descriptor nbd_process_desc =
+	PROC_DESC ( struct nbd_session, process, nbd_tx_step );
+
+
+/**
+ * Process initial negotiation
+ *
+ * @v nbd		NBD session
+ * @v data		Received data
+ * @v len		Length of received data
+ * @ret rc		Return status code
+ */
+static int nbd_rx_neg_init ( struct nbd_session *nbd ) {
+	DBGC2 ( nbd, "NBD %p initial handshake:\n", nbd );
+        DBGC2_HD ( nbd, &nbd->rx_neg_init, sizeof ( nbd->rx_neg_init ) );
+
+	if ( ntohll ( nbd->rx_neg_init.init_magic ) != NBD_INIT_PASSWD ||
+	     ntohll ( nbd->rx_neg_init.opt_magic ) != NBD_OPT_REQ_MAGIC ) {
+		DBGC ( nbd, "NBD %p initial handshake failed (1)\n", nbd );
+		return -EPROTO;
+	}
+
+	nbd->handshake_flags = ntohs ( nbd->rx_neg_init.handshake_flags );
+	if ( ! ( nbd->handshake_flags & NBD_FLAG_FIXED_NEWSTYLE ) ) {
+		DBGC ( nbd, "NBD %p initial handshake failed (2)\n", nbd );
+		return -EPROTO;
+	}
+
+	/* Start transmit an option */
+	nbd_start_tx ( nbd, NBD_TX_NEG_OPT );
+	return 0;
+}
+
+/**
+ * Process reply for NBD_OPT_EXPORT_NAME
+ *
+ * @v nbd		NBD session
+ * @v data		Received data
+ * @v len		Length of received data
+ * @ret rc		Return status code
+ */
+static int nbd_rx_neg_exp_name ( struct nbd_session *nbd ) {
+	DBGC2 ( nbd, "NBD %p export name reply:\n", nbd );
+        DBGC2_HD ( nbd, &nbd->rx_exp_name_reply,
+		   sizeof ( nbd->rx_exp_name_reply ) );
+
+	nbd->export_size = ntohll ( nbd->rx_exp_name_reply.export_size );
+	nbd->trans_flags = ntohs ( nbd->rx_exp_name_reply.trans_flags );
+
+	/* Discard 124 bytes if applicable */
+	if ( ! ( nbd->handshake_flags & NBD_FLAG_NO_ZEROES ) )
+		nbd->discard_len = 124;
```

Comment:
Where does this magic 124 come from?

Reply:
See https://github.com/NetworkBlockDevice/nbd/blob/master/doc/proto.md#newstyle-negotiation
This constant is unnamed in protocol documentation

Reply:
OK